web/flat-web/src/pages/DevicesTestPage/index.tsx

```diff
@@ -0,0 +1,175 @@
+
+    const onDevicesChanged = useCallback((devices: MediaDeviceInfo[]) => {
+        const cameraDevices: MediaDeviceInfo[] = [];
+        const speakerDevices: MediaDeviceInfo[] = [];
+        const microphoneDevices: MediaDeviceInfo[] = [];
+        for (const device of devices) {
+            if (!device.deviceId) {
+                continue;
+            }
+            switch (device.kind) {
+                case "audioinput": {
+                    microphoneDevices.push(device);
+                    break;
+                }
+                case "audiooutput": {
+                    speakerDevices.push(device);
+                    break;
+                }
+                case "videoinput": {
+                    cameraDevices.push(device);
+                    break;
+                }
+            }
+        }
+        setCameraDevices(cameraDevices);
+        setMicrophoneDevices(microphoneDevices);
+        setIsCameraAccessible(cameraDevices.length > 0);
+        setIsMicrophoneAccessible(microphoneDevices.length > 0);
+    }, []);
+
+    useEffect(() => {
+        deviceTest.onVolumeChanged = setVolume;
+        deviceTest.onDevicesChanged = onDevicesChanged;
+        deviceTest.initialize().catch((error: any) => {
+            if (DeviceTest.isPermissionError(error)) {
+                setIsCameraAccessible(false);
+                setIsMicrophoneAccessible(false);
+            }
+        });
+
+        return deviceTest.destroy.bind(deviceTest);
+    }, [deviceTest, onDevicesChanged]);
+
+    useEffect(() => {
+        if (cameraDeviceId) {
+            void deviceTest.setCamera(cameraDeviceId).catch((error: any) => {
+                if (DeviceTest.isPermissionError(error)) {
+                    setIsCameraAccessible(false);
+                }
+            });
+        }
+    }, [cameraDeviceId, deviceTest]);
+
+    useEffect(() => {
+        if (microphoneDeviceId) {
+            void deviceTest.setMicrophone(microphoneDeviceId).catch((error: any) => {
+                if (DeviceTest.isPermissionError(error)) {
+                    setIsMicrophoneAccessible(false);
+                }
+            });
+        }
+    }, [microphoneDeviceId, deviceTest]);
+
+    const refreshDevices = useCallback(() => {
+        refreshDevicesTimer.current = window.setTimeout(() => {
+            refreshDevicesTimer.current = 0;
+            void deviceTest.refreshDevices();
+            refreshDevices();
+        }, 1000);
+    }, [deviceTest]);
+
+    useEffect(() => {
+        if (isCameraAccessible && isMicrophoneAccessible) {
+            return;
+        }
+
+        if (refreshDevicesTimer.current) {
+            return;
+        }
+        refreshDevices();
+    }, [isCameraAccessible, deviceTest, isMicrophoneAccessible, refreshDevices]);
+
+    const closeDeviceTest = (): void => {
+        globalStore.switchDeviceTest();
+        console.log("is open device test", globalStore.isCloseDeviceTest);
```

Comment:
`closeDeviceTest` / `switchDeviceTest` /  `is open device`?

Variable names are problematic